credential/cargo-credential-macos-keychain/src/lib.rs

```diff
@@ -0,0 +1,77 @@
+//! Cargo registry macos keychain credential process.
+
+#[cfg(target_os = "macos")]
+mod macos {
+    use cargo_credential::{
+        read_token, Action, CacheControl, Credential, CredentialResponse, Error, RegistryInfo,
+    };
+    use security_framework::os::macos::keychain::SecKeychain;
+
+    pub struct MacKeychain;
+
+    /// The account name is not used.
+    const ACCOUNT: &'static str = "";
+
+    fn registry(index_url: &str) -> String {
+        format!("cargo-registry:{}", index_url)
+    }
+
+    fn to_credential_error(e: security_framework::base::Error) -> Error {
+        Error::Other(format!("security framework: {e}"))
+    }
+
+    impl Credential for MacKeychain {
+        fn perform(
+            &self,
+            reg: &RegistryInfo<'_>,
+            action: &Action<'_>,
+            _args: &[&str],
+        ) -> Result<CredentialResponse, Error> {
+            let keychain = SecKeychain::default().unwrap();
+            match action {
+                Action::Get(_) => {
+                    let service_name = registry(reg.index_url);
+                    let (pass, _item) = keychain
+                        .find_generic_password(&service_name, ACCOUNT)
+                        .map_err(to_credential_error)?;
+                    let token = String::from_utf8(pass.as_ref().to_vec())
+                        .map_err(|_| Error::Other("failed to convert token to UTF8".to_string()))?;
+                    Ok(CredentialResponse::Get {
+                        token: token.into(),
+                        cache: CacheControl::Session,
+                        operation_independent: true,
+                    })
+                }
+                Action::Login(options) => {
+                    let service_name = registry(reg.name.unwrap_or(reg.index_url));
```

Comment:
There is something wrong here, though I don't think it was introduced in this PR (I'm not sure, but looks like this regressed in #10592).  This is failing for me because `login` is using a different service name than `get`. Is this something you can fix? If not, I can look at it later. It's ok to do that later if needed.

Reply:
Yes, it looks like I broke it in #10592 😢. I'll fix it and try to borrow a Mac to validate it.